App/Services/HostnameSuffixGetter.cs

```diff
@@ -0,0 +1,145 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Coder.Desktop.App.Models;
+using Coder.Desktop.CoderSdk.Coder;
+using Coder.Desktop.Vpn.Utilities;
+using Microsoft.Extensions.Logging;
+
+namespace Coder.Desktop.App.Services;
+
+public interface IHostnameSuffixGetter
+{
+    public event EventHandler<string> SuffixChanged;
+
+    public string GetCachedSuffix();
+}
+
+public class HostnameSuffixGetter : IHostnameSuffixGetter
+{
+    private const string DefaultSuffix = ".coder";
+
+    private readonly ICredentialManager _credentialManager;
+    private readonly ICoderApiClientFactory _clientFactory;
+    private readonly ILogger<HostnameSuffixGetter> _logger;
+
+    // _lock protects all private (non-readonly) values
+    private readonly RaiiSemaphoreSlim _lock = new(1, 1);
+    private string _domainSuffix = DefaultSuffix;
+    private bool _dirty = false;
+    private bool _getInProgress = false;
+    private CredentialModel _credentialModel = new() { State = CredentialState.Invalid };
+
+    public event EventHandler<string>? SuffixChanged;
+
+    public HostnameSuffixGetter(ICredentialManager credentialManager, ICoderApiClientFactory apiClientFactory,
+        ILogger<HostnameSuffixGetter> logger)
+    {
+        _credentialManager = credentialManager;
+        _clientFactory = apiClientFactory;
+        _logger = logger;
+        credentialManager.CredentialsChanged += HandleCredentialsChanged;
+        HandleCredentialsChanged(this, _credentialManager.GetCachedCredentials());
+    }
+
+    ~HostnameSuffixGetter()
+    {
+        _credentialManager.CredentialsChanged -= HandleCredentialsChanged;
+    }
+
+    private void HandleCredentialsChanged(object? sender, CredentialModel credentials)
+    {
+        using var _ = _lock.Lock();
+        _logger.LogDebug("credentials updated with state {state}", credentials.State);
+        _credentialModel = credentials;
+        if (credentials.State != CredentialState.Valid) return;
+
+        _dirty = true;
+        if (!_getInProgress)
+        {
+            _getInProgress = true;
+            Task.Run(Refresh).ContinueWith(MaybeRefreshAgain);
+        }
+    }
+
+    private async Task Refresh()
```

Comment:
This should probably just take the credentials model as an argument.

Reply:
No, we want this to operate on the "latest" credentials since we're lock anyway to set `_dirty`.